[CDAP-16286] Added FLL for database source and sink plugins

vravish opened the pull request.

vravish committed:
[CDAP-16286] Added FLL for database source and sink plugins

vravish committed:
[CDAP-16286] Added FLL for database source and sink plugins

yaojiefeng commented on database-plugins/src/main/java/io/cdap/plugin/db/batch/sink/DBSink.java:
```diff
@@ -115,6 +117,9 @@ public void prepareRun(BatchSinkContext context) {
     context.addOutput(Output.of(dbSinkConfig.referenceName, new DBOutputFormatProvider(dbSinkConfig, driverClass)));
+
+    recordLineage(context, dbSinkConfig.referenceName, Schema.of(Schema.Type.STRING),
+        Arrays.asList(dbSinkConfig.columns.split(",")));
```
The field must be derived from the schema

yaojiefeng replied:
Also the indentation is incorrect. This line should align with the context like: 
 ```
recordLineage(context, dbSinkConfig.referenceName, Schema.of(Schema.Type.STRING), 
               outputFields)
```

yaojiefeng commented on database-plugins/src/main/java/io/cdap/plugin/db/batch/sink/DBSink.java:
```diff
@@ -115,6 +117,9 @@ public void prepareRun(BatchSinkContext context) {
     }
     context.addOutput(Output.of(dbSinkConfig.referenceName, new DBOutputFormatProvider(dbSinkConfig, driverClass)));
+
+    recordLineage(context, dbSinkConfig.referenceName, Schema.of(Schema.Type.STRING),
```
Why do you construct a new Schema here? We should NEVER construct the schema. You can get the schema using `context.getInputSchema()`. And then fields can be derived like https://github.com/cdapio/hydrator-plugins/blob/develop/format-common/src/main/java/io/cdap/plugin/format/plugin/AbstractFileSink.java#L96. Please do take a look what the context can give you since that is essential about the fll

yaojiefeng commented on database-plugins/src/main/java/io/cdap/plugin/db/batch/sink/DBSink.java:
```diff
@@ -190,6 +195,17 @@ private void setResultSetMetadata() throws Exception {
   }
 
+  private void recordLineage(BatchSinkContext context,
+                             String outputName,
```
nit - you can put some of the variables in the same line

yaojiefeng commented on database-plugins/src/main/java/io/cdap/plugin/db/batch/source/DBSource.java:
```diff
@@ -357,4 +358,16 @@ private Schema getSchema(FailureCollector collector) {
+
+    if (schema != null && schema.getFields() != null) {
+      lineageRecorder.recordRead("Read", "Read from DB.",
+          schema.getFields().stream().map(Schema.Field::getName).collect(
```
nit - the indentation is wrong. It should align with "Read" in the previous line

yaojiefeng commented on database-plugins/src/main/java/io/cdap/plugin/db/batch/source/DBSource.java:
```diff
@@ -357,4 +358,16 @@ private Schema getSchema(FailureCollector collector) {
   }
+
+  private void emitLineage(BatchSourceContext context) {
+    Schema schema = sourceConfig.getSchema(context.getFailureCollector());
```
config.getSchema() can return null, if it is null, you should get it from `context.getOutputSchema`

yaojiefeng reviewed (commented):
Please do get yourself familiar with what the context class can give you

vravish committed:
[CDAP-16286] Added FLL for database source and sink plugins

vravish committed:
[CDAP-16286] Added FLL for database source and sink plugins

vravish commented:
Resolved the comments from Yaojie earlier.

yaojiefeng commented on database-plugins/src/main/java/io/cdap/plugin/db/batch/sink/DBSink.java:
```diff
@@ -115,6 +117,10 @@ public void prepareRun(BatchSinkContext context) {
+
+    Schema schema = context.getInputSchema();
+    recordLineage(context, dbSinkConfig.referenceName, schema,
+                  schema.getFields().stream().map(Schema.Field::getName).collect(Collectors.toList()));
```
schema can be null, your IDE should warn you about possible NPE, please do not ignore them

vravish replied:
Ack.

yaojiefeng commented on database-plugins/src/main/java/io/cdap/plugin/db/batch/source/DBSource.java:
```diff
@@ -357,4 +358,17 @@ private Schema getSchema(FailureCollector collector) {
+
+  private void emitLineage(BatchSourceContext context) {
+    Schema schema = sourceConfig.getSchema(context.getFailureCollector());
+    if (schema == null)
```
should always have `{` after the if condition, the checkstyle will fail, make sure you run a `mvn clean package -DskipTests` before you commit the code

vravish replied:
Ack.

vravish committed:
[CDAP-16286] Added FLL for database source and sink plugins

vravish committed:
[CDAP-16286] Added FLL for database source and sink plugins

vravish committed:
[CDAP-16286] Added FLL for database source and sink plugins

vravish committed:
[CDAP-16286] Added FLL for database source and sink plugins

yaojiefeng reviewed (commented):
Please remove the submodule changes, other changes LGTM

albertshau committed:
CDAP-16002 fix bug where header is emitted multiple times

When multiple files with headers are being read, the header will
be emitted once per file within the same RecordReader. This causes
Wrangler's parse-as-csv with header to fail when it reaches the
header for the 2nd file.

To fix this, ignore the header for everything except for the first
file in the split. This is only done when the hidden copyHeader
variable is set, as we do not actually want to filter out the first
row for files that do not have a special header.

albertshau committed:
update condition plugins commit to one that exists

albertshau committed:
Merge pull request #1023 from cdapio/fix-conditions-reference

update condition plugins commit to one that exists

albertshau committed:
Merge pull request #1022 from cdapio/bugfix/CDAP-16002-fix-multifile-headers

CDAP-16002 fix bug where header is emitted multiple times

vravish committed:
Merge branch 'feature/CDAP-16286-fll-to-database-plugins' of https://github.com/cdapio/hydrator-plugins into feature/CDAP-16286-fll-to-database-plugins

vravish committed:
CDAP-16286

vravish committed:
[CDAP-16286] Added FLL for database source and sink plugins
